Experimental fit tutorial (#463)
dynamiqs/dynamiqs: docs/tutorials/experimental-fits.md

> # Experimental data fits

Reviewer: ```suggestion
# Fitting a cat qubit experiment with real-world data
```

> Dynamiqs enables rapid simulations and facilitates automatic gradient computation, making it highly suitable for data fitting tasks. This tutorial will guide you through the process of simulating and fitting real-world experimental data, step by step.

Reviewer: ```suggestion
dynamiqs enables fast simulations of quantum systems with automatic gradient computation, making it highly suitable for data fitting tasks. This tutorial will guide you through the process of simulating and fitting real-world experimental data, step by step.
```

## Context of the experiment

!!! note Info

>     This part introduces the kind of data and process that we will fit in the notebook. If you are only interested in the code you can skip it.

Reviewer: ```suggestion
    This section introduces the kind of data and process that we will fit in the notebook. If you are only interested in the code you can skip the next section.
```

!!! example "Diving deep 🤿"

>     You can discover further details about this experiment in the original research paper from which this data was sourced [^1]

Reviewer: I think this is a bit heavy/insistent: you should definitely cite the paper but probably within the next paragraph.

> The data we will work with are deflation curves. Our system is composed of a high lifetime cavity - the memory -  coupled to a lossy one - the buffer - in vacuum state. We denote $a$ (respectively $b$) the memory (respectively buffer) annihilation operator. The memory and the buffer are coupled through a two to one photon exchange mechanism $g_2 a^2 b^\dagger + h.c.$. The buffer is a highly-dissipative mode, that we model with the dissipator $D[\sqrt{\kappa_b} b]$. Both cavities have a finite temperature, expressed in thermal population.

Reviewer: Start with a short but higher-view description of the experiment (in which you can cite the paper). This is a bit rough to start with "The data we will work with are deflation curves" which doesn't mean anything to the regular physicist.

Author: >Our system is composed of a high lifetime cavity - the memory -  coupled to a lossy one - the buffer - in vacuum state. We denote $a$ (respectively $b$) the memory (respectively buffer) annihilation operator. The memory and the buffer are coupled through a two to one photon exchange mechanism $g_2 a^2 b^\dagger + h.c.$. The buffer is a highly-dissipative mode, that we model with the dissipator $D[\sqrt{\kappa_b} b]$. Both cavities have a finite temperature, expressed in thermal population. For more details, please see [^1].

> At the beginning of the experiment the memory is populated with a coherent state and the buffer is in thermal vacuum. We activate the two photons exchange mechanism. Pairs of photons leave the memory and are converted into a single buffer photon, that is then dissipated into the environment. The memory photon number $a^\dagger a$ is measured at different times and saved into the `data` variable. Our goal is to infer $g_2$ from the data.

> At the beginning of the experiment the memory is populated with a coherent state and the buffer is in thermal vacuum. We enable for a period of two photon exchange mechanism. During this period, pairs of photons leave the memory and are converted into a single buffer photon, that is then dissipated into the environment. The memory photon number $a^\dagger a$ is measured at different times and saved into the `data` variable. Our goal is to infer $g_2$ from the data.

Reviewer: "We enable for a period of two photon exchange mechanism"

Not sure what this means. Do you mean you turn on the two-photon exchange mechanism?

Author: Changed to 
> At the beginning of the experiment the memory is populated with a coherent state and the buffer is in thermal vacuum. We activate the two photons exchange mechanism. Pairs of photons leave the memory and are converted into a single buffer photon, that is then dissipated into the environment. The memory photon number $a^\dagger a$ is measured at different times and saved into the `data` variable. Our goal is to infer $g_2$ from the data.

![Pulse sequence](/figs-docs/g2-pulse-sequence.png)
_Pulse sequence of the experiment, taken from [^1]_

[^1]:  [Quantum control of a cat-qubit with bit-flip times exceeding ten seconds, arxiv:2307.06617](https://arxiv.org/pdf/2307.06617.pdf)

> ## Step by step solution

Reviewer: ```suggestion
## Fitting real-world data, step by step
```

> 

Reviewer: Can you add a short introduction to what you will be doing in this section?

Author: > In this notebook, we will load the experimental data (or simulate some), encode the two photon exchange model in dynamiqs, simulate the system evolution and measure the number of photons at different instants. We will then use gradient descent to fit the $g_2$ parameter to the data.

We start by importing all necessary libraries:

```python
import dynamiqs as dq
import jax
import jax.numpy as jnp
import matplotlib.pyplot as plt
import numpy as np
from tqdm import tqdm
```

Declare useful helpers to handle units:

```python
MHz = 2 * np.pi
kHz = 2 * np.pi * 1e-3
us = 1.0
ns = 1.0e-3
```

Define physical values that describe our system (taken from [^1]):

```python
thermal_a, thermal_b = 0.1, 0.011
kappa_a = 9.3e-3 * kHz
kappa_b = 2.6 * MHz
```

Instanciate our bosonic annihilation operators:

> ```python

Reviewer: If you only have single-line comments about each of these things, I would suggest putting them all in a single code block, with python comments above. It's easier to read and to follow.

Na, Nb = 15, 7
a, b = dq.destroy(Na, Nb))
```

### Loading the data

> We provide real experimental data but also leave the possibility to simply simulate them, you can choose the one that suits you best

Reviewer: ```suggestion
We now initialize the experimental data with which we will work. There are two possible options to do so: either load the actual experimental data from [^1]_, or generate synthetic data from a noisy master equation simulation.
```

#### Option 1: Real world data

> Before running the code, you can download the data at this link (TODO: add link)

Reviewer: I guess the data should be put somewhere on the github (probably in the same folder), so people don't have to open a link to some scary website.

Author: Indeed, but the files are nowhere on github yet

> Load the data

Reviewer: Same comment as before: either say a bit more about what you're doing, or put these comments inside the code-block. Right now it's a bit rough around the edges.

```python
def load(file):
    x = np.loadtxt(file)
    x = jnp.array(x)
    return x

data = load("short.npy")
time = load("time_short.npy")
alpha2 = data[0]
```

and plot them

```python
plt.figure(figsize=(10, 5))

plt.plot(time, data, "+")
plt.grid()
plt.xlabel("Time [us]")
plt.ylabel("Photon number")
renderfig('g2-real-data')
```

% invisible-code-block: python

> % exp_time, exp_data, exp_alpha2 = time, data, alpha2

Reviewer: Since this doesn't work yet, either we first fix it in a separate PR, or we remove this altogether (for now).

Author: The docs won't run without this

Author: Yes let's fix it 👍

![Loaded data](/figs-docs/g2-real-data.png)

#### Option 2: Synthetic data

Create the synthetic data 

> ```python

Reviewer: This whole block/section is lacking comments or descriptions.

We define the simulation code

```python
def simulate(g2):
    b0 = dq.unit(dq.fock(Nb, 0) + thermal_b ** 0.5 * dq.fock(Nb, 1))
    rho0 = jnp.stack([dq.tensor(dq.coherent(Na, alpha), b0) for alpha in alpha2 ** 0.5])

    H = g2 * a @ a @ dq.dag(b)
    H += dq.dag(H)

    La_down = np.sqrt(kappa_a * (1 + thermal_a)) * a
    La_up   = np.sqrt(kappa_a * thermal_a) * dq.dag(a)
    Lb_down = np.sqrt(kappa_b * (1 + thermal_b)) * b
    Lb_up   = np.sqrt(kappa_b * thermal_b) * dq.dag(b)

    result = dq.mesolve(
        H, [Lb_down, Lb_up, La_down, La_up],
        rho0, time,
        exp_ops=[dq.dag(a) @ a],
    )

    return result
```

We choose an initial guess that is reasonable with respect to the data we have. We converge to a steady state in roughly $1.0\mu$s so $g_2$ should be of the order of $1$MHz. Let's take $0.1$MHz to have some margin.
```python
g2 = jnp.array(0.1 * MHz)
result = simulate(g2)

plt.figure(figsize=(10, 5))

time_np = time
for i in range(len(alpha2)):
    plt.plot(time_np, data[:, i].real, f"+C{i}")
    plt.plot(time_np, result.expects[i, 0].real, f"C{i}")
plt.grid()
plt.xlabel("Time [us]")
plt.ylabel("Photon number")
renderfig('g2-initial-guess')
```

> ![g2 initial guess](/figs-docs/g2-initial-guess.png)

Reviewer: This figure looks a bit silly IMO. Maybe taking $g_2$ to be much larger (say 5 MHz) would be better?

Author: It makes sense to show that the fit is bad in a first place. At 5MHz you have a batch of oscillations, it feels like a deliberately bad choice

Author: ![Uploading image.png…]()

Reviewer: I think it looks better, we clearly see it's not the right value no?

Author: We also see it with the low value, and it's more complicated  to me. It makes more sense to start from lower values because it entails less non linearities in you system so it's easier to understand

> % skip: start

Reviewer: Same comment as before: either we fix it or we remove it.

The optimization loop is defined as follow
```python
def cost(g2, data):
    res = simulate(g2)
    return jnp.sum((data - res.expects[:, 0].real.T) ** 2)

g2 = jnp.array(0.1 * MHz)
costs = []
g2s = [g2]
for _ in tqdm(range(40)):
    val, grad = jax.value_and_grad(cost)(g2, data)
    grad = jnp.clip(grad, -100, 100)
    g2 = g2 - 0.01 * grad
    costs.append(val)
    g2s.append(g2)
costs.append(cost(g2, data))

costs = np.array(costs)
g2s = np.array(g2s)
```
we use a plain gradient descent algorithm for simplicity but one could imagine using any other algorithm, such as the ones provided in [Optax](https://github.com/google-deepmind/optax).

!!! note Optimisation time
    The full optimisation runs in less than one minute on a NVIDIA GeForce RTX 4090 and around 5 minutes on a M1 Mac.

> We can plot the loss to ensure we converged correctly

Reviewer: Where is this plot?

Author: Fixed

```python
plt.figure(figsize=(10, 5))
plt.subplot(121)
plt.plot(costs)
plt.grid()
plt.xlabel("iteration")
plt.ylabel("loss")
plt.yscale("log")

plt.subplot(122)
plt.plot(g2s / MHz)
plt.grid()
plt.xlabel("iteration")
plt.xlabel("g2 [MHz]")
```
![Fit result](/figs-docs/g2-fit-result.png)

Finally we plot the fit and compare with the data
```python
result = simulate(g2)
plt.figure(figsize=(10, 5))

time_np = time
for i in range(len(alpha2)):
    plt.plot(time_np, data[:, i].real, f"+C{i}")
    plt.plot(time_np, result.expects[i, 0].real, f"C{i}")
plt.grid()
plt.xlabel("Time [us]")
plt.ylabel("Photon number")

print(g2 / MHz)
```

!!! example "Final note"
    $g_2$ value differs from the one reported in
    [arxiv:2307.06617](https://arxiv.org/pdf/2307.06617.pdf)
    by approximately 15%. This discrepancy is attributed to the unavailability 
    of Dynamiqs for performing fits at the time of writing the article. 
    Consequently, the authors were unable to employ gradient descent for their

>     fit and had to resort to alternative methods.

Reviewer: I would remove this altogether. You're just giving the baton pour te faire battre here.

> ## Full code

Reviewer: Ideally we'll have a link at some point for this, but for now it's fine I think.
